### PR TITLE
Toloka kit and ClearML integration

### PR DESCRIPTION
**Pull request by erezalg**
Integrating clearml into toloka kit

## Description
Added 2 methods, register_to_clearml() that takes a dataframe from toloka and registers to clearml and get_clearml_dataset() that gets data stored in clearml

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation and examples improvement (changes affected documentation and/or examples)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

I hereby agree to the terms of the CLA available at: [link](https://yandex.ru/legal/cla/?lang=en).

### CODE REVIEW

File: src/__init__.py [hunk 1]
```diff
@@ -17,3 +17,10 @@
     __all__.append('autoquality')
 except ImportError:
     pass
+
+
+try:
+    from src.integrations import clearml  # noqa: F401
+    __all__.append('clearml')
+except ImportError:
+    pass
```
Comment (alexdrydew) [inline src/__init__.py:26]: Please add an empty line at the end of the file

File: src/integrations/clearml/__init__.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+url_regex = re.compile(
+        r'^(?:http|ftp)s?://'
+        r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'
+        r'localhost|'
+        r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'
+        r'(?::\d+)?'
+        r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+
+
+def get_clearml_dataset(
```
Comment (alexdrydew) [inline src/integrations/clearml/__init__.py:25]: I believe that this function can be omitted (at least in current form) as it does not provide any value over `Dataset` API. Unlike this function `Dataset.get(...).get_local_copy()` call supports all current and future parameters of the `Dataset.get ` and performs superior validation of the input parameters. Correct me if I am missing some functionality.

File: src/integrations/clearml/__init__.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    toloka_dataset_path: Optional[str] = None,
+    external_url_columns: Optional[List[str]] = None,
+    dataset_project: Optional[str] = None,
+    dataset_name: Optional[str] = None,
+    parent_dataset_ids: Optional[List[str]] = None,
+    verbose: bool = False
+) -> None:
+    """Registers a new Dataset to ClearML containing files annotated by Toloka
+    Args:
+        toloka_df: pandas Dataframe containing data from Toloka
```
Comment (alexdrydew) [inline src/integrations/clearml/__init__.py:68]: Please describe ways of obtaining this data
```suggestion
        toloka_df: pandas Dataframe containing data from Toloka. Can be obtained using `TolokaClient.get_assignments_df` or using web UI.
```

File: src/integrations/clearml/__init__.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        )
+    else:
+        parent = Dataset.get(dataset_id=parent_dataset_ids)
+        dataset = Dataset.create(
+            dataset_project=parent.project if not None else parent.project,
+            dataset_name=parent.name if not None else parent.name,
+            parent_datasets=[parent_dataset_ids]
+        )
+
+    toloka_csv_file = toloka_dataset_path if toloka_dataset_path is not None else '{}/toloka_ds_temp.csv'.format(tempfile.gettempdir())
```
Comment (alexdrydew) [inline src/integrations/clearml/__init__.py:105]: Please use f-string here

File: src/integrations/clearml/__init__.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    toloka_df.to_csv(toloka_csv_file, index=False)
+    dataset.add_files(path=toloka_csv_file, verbose=verbose)
+
+    if external_url_columns is not None:
+        for column_name in toloka_df.columns:
+            # if column_name.startswith('INPUT:') or column_name.startswith('GOLDEN:') or column_name.startswith('OUTPUT:'):
```
Comment (alexdrydew) [inline src/integrations/clearml/__init__.py:111]: Please remove the commented code

File: src/integrations/clearml/__init__.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            if column_name in external_url_columns:
+                sample_value = toloka_df[column_name].iloc[0]
```
Comment (alexdrydew) [inline src/integrations/clearml/__init__.py:113]: This check is unfriendly to datasets with optional values (e.g. some answers may contain url while some will not) as the first row may happen to be empty. I suggest to restructure the method: first, download assignments from Toloka, then validate if every row in external_url_columns is either URL or empty. Then continue from line 89 but this time all non-empty external column values should be added to the dataset unconditionally

File: src/integrations/clearml/__init__.py (hunk 2 above)
Comment (alexdrydew) [inline src/integrations/clearml/__init__.py:25]: Please add unit test for this method (you can mock `Dataset` class)

File: setup.py [hunk 7]
```diff
@@ -24,6 +24,7 @@
     's3': ['boto3 >= 1.4.7'],
     'zookeeper': ['kazoo >= 2.6.1'],
     'jupyter-metrics': ['plotly', 'ipyplot', 'jupyter-dash'],
+    'clearml': ['clearml', 'pandas']
 }
 EXTRAS_REQUIRE['all'] = sum((deps for env, deps in EXTRAS_REQUIRE.items() if env != 'dev'), [])
```
Comment (alexdrydew) [inline setup.py:29]: Let's not include integrations in `all` variable: I suggest to create separate `integrations` variable with all future integrations

File: src/__init__.py [hunk 8]
```diff
@@ -17,3 +17,10 @@
     __all__.append('autoquality')
 except ImportError:
     pass
+
+
+try:
+    from src.integrations import clearml  # noqa: F401
```
Comment (alexdrydew) [inline src/__init__.py:23]: Please use relative import here
```suggestion
    from .integrations import clearml  # noqa: F401
```

Comment (alexdrydew) [inline src/__init__.py:23]: Please include test for this conditional import: you can use this test as a reference:
https://github.com/Toloka/toloka-kit/blob/main/tests/import_tests/test_autoquality.py

Also please create a separate environment in tox configuration and run this test separately:
https://github.com/Toloka/toloka-kit/blob/dfa2bb6b896670de9931ffebf978981a2c6e377a/tox.ini#L4
https://github.com/Toloka/toloka-kit/blob/dfa2bb6b896670de9931ffebf978981a2c6e377a/tox.ini#L33

File: src/integrations/clearml/__init__.py (hunk 2 above)
Comment (DrhF) [reply src/integrations/clearml/__init__.py:25]: I believe this function may be useful when one want to label data that is stored in ClearML Dataset

Comment (DrhF) [reply src/integrations/clearml/__init__.py:25]: > I believe that this function can be omitted (at least in current form) as it does not provide any value over `Dataset` API. Unlike this function `Dataset.get(...).get_local_copy()` call supports all current and future parameters of the `Dataset.get ` and performs superior validation of the input parameters. Correct me if I am missing some functionality.

I believe this function may be useful when one want to label data that is stored in ClearML Dataset

**Pull request closed without merging**